pallets/poa/src/lib.rs

```diff
@@ -0,0 +1,467 @@
+                    if removed_queued > 0 {
+                        queued_validator_set_changed = true;
+                    }
+                }
+            }
+
+            let max_validators = T::MaxActiveValidators::get() as usize;
+
+            // TODO: Remove debugging variable below
+            let mut count_added = 0u32;
+
+            // Make any queued validators active.
+            while (active_validators.len() < max_validators) && (validators_to_add.len() > 0) {
+                active_validator_set_changed = true;
+                queued_validator_set_changed = true;
+                active_validators.push(validators_to_add.remove(0));
+                count_added += 1;
+            }
+
+            // Only write if queued_validator_set_changed
+            if queued_validator_set_changed {
+                <QueuedValidators<T>>::put(validators_to_add);
+            }
+
+            ForceSessionChange::put(false);
+
+            let active_validator_count = active_validators.len() as u8;
+            if active_validator_set_changed {
+                print("Active validator set changed, rotating session");
+                print(count_added);
+                print(count_removed);
+                <ActiveValidators<T>>::put(active_validators);
+            }
+            (active_validator_set_changed, active_validator_count)
+        } else {
+            (false, active_validators.len() as u8)
+        }
+    }
+
+    /// Set next epoch duration such that it is >= `MinEpochLength` and also a multiple of the
+    /// number of active validators
+    fn set_current_epoch_end(current_slot_no: u64, active_validator_count: u8) {
+        let min_epoch_len = T::MinEpochLength::get();
+        let active_validator_count = active_validator_count as u64;
+        let rem = min_epoch_len % active_validator_count;
+        let epoch_len = if rem == 0 {
+            min_epoch_len
+        } else {
+            min_epoch_len + active_validator_count - rem
+        };
+        let epoch_ends_at = current_slot_no + epoch_len;
+        print("epoch ends at");
+        print(epoch_ends_at);
+        EpochEndsAt::put(epoch_ends_at);
+    }
+
+    /// Swap a validator account from active validators. Swap out `old_validator_id` for `new_validator_id`.
+    /// Expects the active validator set to contain `old_validator_id`. This is ensured by the extrinsic.
+    fn swap(old_validator_id: T::AccountId, new_validator_id: T::AccountId) -> u8 {
+        let mut active_validators = Self::active_validators();
+        let count = active_validators.len() as u8;
+        for (i, v) in active_validators.iter().enumerate() {
+            if *v == old_validator_id {
+                active_validators[i] = new_validator_id;
+                break;
+            }
+        }
+        <ActiveValidators<T>>::put(active_validators);
+        count
+    }
+
+    fn current_slot_no() -> Option<u64> {
```

Comment:
Firstly, and least importantly, `Option` implements `Try`, taking advantage of that will make this function more readable:

```rust
let digest = <system::Module<T>>::digest();
let first = digest.logs().first()?;
let pre_run = first.as_pre_runtime()?;
Some(u64::decode(&mut &pre_run.1[..]).unwrap())
```

How do you know that slot number is stored in a `PreRuntime` variant in first log item in the block header? Who sets it? Also how do you know it's encoded as a u64?

If the log has no items or the first item is not a `PreRuntime`, `should_end_session` below returns false. Isn't that answer potentially incorrect? It should panic instead of silently ignoring the error, right?

Reply:
tags:
- readability-surface
  - see suggestion
- documentation
  1. The assumptions this code makes are explicitly stated, which is good. Next step is argue for the validity of the assumptions. These arguments may take the form of links to substrate documentation, debug_assertions and prose.
  2. It's not clear *why* the current slot would be inaccessible. A description of this should be in the doc comment.
- bugs
  1. During review, It's unclear to me whether this function will behave correctly. Documentation may clarify.

Reply:
This code has changed [in the other branch](https://github.com/docknetwork/dock-substrate/blob/emission/pallets/poa/src/lib.rs#L659) and not returns None and does not have the `?`. 

> How do you know that slot number is stored in a PreRuntime

[Here](https://github.com/paritytech/substrate/blob/v2.0.0-rc4/client/consensus/slots/src/lib.rs#L238) is the digest item created and added to logs and the slot no is u64. The PreRuntime is created in `pre_digest_data` implementation of aura as shown [here](https://github.com/paritytech/substrate/blob/v2.0.0-rc4/client/consensus/aura/src/lib.rs#L258). Will mention it as code comment. There is another way to do as well using `as_aura_pre_digest` but that will cause me to add another build dependency as well (which is not wrong since its supposed to be used with Aura). 

> Who sets it?

The validator set is [here](https://github.com/paritytech/substrate/blob/v2.0.0-rc4/client/consensus/slots/src/lib.rs#L241)

> If the log has no items or the first item is not a PreRuntime, should_end_session below returns false. Isn't that answer potentially incorrect? It should panic instead of silently ignoring the error, right?

The block's digest is not always available. But whenever I use it (for non-debugging purposes), I unwrap and I have not seen it panic yet so I assume its not an issue.